### PR TITLE
Cache last reported current state

### PR DESCRIPTION
**Pull request by cywang117**
Whenever the Supervisor reports current state, it diffs the current state
with its last reported current state. However, when the Supervisor starts
up, there is no last reported state, since that last report is stored in
process memory. Caching the last report in a location that survives
Supervisor restarts will reduce the current report bandwidth used on startup.

Change-type: patch
Signed-off-by: Christina Ying Wang <christina@balena.io>

### CODE REVIEW

**Review by Page- — COMMENTED** [review]
I think it makes more sense to cache the target state when we get a new one rather than on every supervisor exit, this helps for a few reasons:
1. If the supervisor exists unexpectedly (eg power pulled) we still have the cache
2. We only write when the cache has actually changed, and it will be at a time when we're almost certainly doing writes anyway in order to attempt to apply the target state so there shouldn't be any real worries about increased writes (since it won't be particularly significant)
3. It's probably simpler/more robust than hooking into process exiting if we don't need to

**Comment by cywang117** [comment]
@Page- Thanks for the feedback. The original implementation was caching last reported current state after every report. I'd discussed this with @pipex previously and we were concerned about  the increased I/O that would involve on the device with every cache operation. Leading to the current implementation of it. I'm going to run the draft Supervisor on an RPi 0 to see if the memory impact is significant and will report back. For the record, I'm okay with either implementation.

**Comment by pipex** [comment]
This looks good @cywang117  i don't really have any observations. Merge when you are ready with testing

**Comment by cywang117** [comment]
@balena-ci I self-certify!

File: src/api-binder/report.ts [hunk 1]
```diff
@@ -128,6 +138,43 @@ async function reportCurrentState(opts: StateReportOpts, uuid: string) {
...
+		// Touch the file, which will create it if it doesn't exist
+		await touch(CACHE_PATH);
+
+		// Get last reported current state
+		const rawStateCache = await readFile(CACHE_PATH, 'utf-8');
+		const state = JSON.parse(rawStateCache);
+
+		// Return current state cache if valid
+		if (!DeviceState.is(state)) {
+			throw new Error();
```
Comment (thgreasi) [inline src/api-binder/report.ts:166]: Shouldn't it delete the cache file if it's invalid rather than throwing and retrying forever?

**Pull request merged**